### ISSUE TITLE
MAUI WebView doesn't allow to web session to inspected in Safari developer tools (Mac Catalyst)

### ISSUE THREAD

**Issue by alexeystrakh** [dotnet/maui#7706]
### Description

WKWebView Safari developer tools (inspect option) are not available in a MAUI Mac app. If the same app is created using native tooling Xcode + Swift with no extra configuration/entitlements - it works (there is a visible session to inspect in Safari dev tools). If a new MAUI app is created with just a WebView in it (which created WKMauiWebView, which is essentially the same as WKWebView), Safari shows no debuggable sessions available.

Other scenarios tests - the same result with no option to debug the session: 
- BlazorWebView
- Plain MacCatalyst WKWebView
- MAUI WebView with custom handler with `developerExtrasEnabled` set

The source for all foud projects is in the following repository:
https://github.com/xamcat/maui-wkwebview-inspect

### Steps to Reproduce

1. Create new MAUI App
2. Register WebView on the main page

```
 <WebView Source="https://www.apple.com"
             VerticalOptions="FillAndExpand"
             HorizontalOptions="FillAndExpand" />
```

3. Run on Mac with `dotnet build -t:Run -f net6.0-maccatalyst`
4. Open Safari and navigate to `Develop-> <MacName>`

Result: `No inspectable application` message
Expected: Web session created by WebView (WKWebView) is visible in debug tools

<img width="1864" alt="safari-debug-native-mac-app" src="https://user-images.githubusercontent.com/3697084/171707288-79639bad-9190-4374-9d6b-27985ff35d86.png">
<img width="2282" alt="safari-debug-maui-mac-app" src="https://user-images.githubusercontent.com/3697084/171707387-99f8b92f-2ffc-4bd5-82e1-b4e349fd8418.png">
<img width="2263" alt="safari-debug-maccat-mac-app" src="https://user-images.githubusercontent.com/3697084/171707462-5b1b5fb7-b942-4d0b-a979-1131d1b9e7db.png">
<img width="1140" alt="safari-debug-blazor-mac-app" src="https://user-images.githubusercontent.com/3697084/171707741-55b19186-312c-4ad0-acb8-00df149e5cc6.png">

### Version with bug

6.0 (current)

### Last version that worked well

Release Candidate 3 (current)

### Affected platforms

macOS

### Affected platform versions

maccatalyst 13.1

### Did you find any workaround?

No workaround has been found.

**Comment by Eilon** [comment]
We tried everything we could to get this working with BlazorWebView on Mac, and we just couldn't get it to work. So it's probably the same problem as the regular .NET MAUI WebView, but we don't know what that is. We got Windows (WebView2), Android, and iOS all working with Dev Tools, but not macOS. But it's interesting to hear that an app built in Xcode has this working, so maybe we can compare what that does to what MAUI is doing, and figure out the difference.

**Comment by javiercn** [comment]
@Eilon do you think we might be missing an entitlement in the app manifest?

**Comment by Eilon** [comment]
@javiercn I'm not sure. I didn't see any entitlements in the Xcode app, either, unless it's somehow auto-generated in some cases? (Maybe let's say in debug mode?) Or maybe I missed where it is? Also, we did try using entitlements to enable F12 tools but it never worked, and various Internet people had also encountered the same thing, so we always ended up at a dead end.

**Comment by alexeystrakh** [comment]
I found some confirmations that the entitlement is needed ([get-task-allow](https://stackoverflow.com/questions/45827865/how-can-i-use-safari-to-debug-a-wkwebview-in-my-mac-app)) to have the Safari inspect tools work, but also confirmed that Xcode doesn't include any entitlements for a Cocoa app and it works for Xcode. 

May be Xcode includes this entitlements for debug builds automatically without asking us? [Found here](https://developer.apple.com/forums/thread/118415)

> `get-task-allow` is typically added by Xcode automatically for development builds

**Comment by TanayParikh** [comment]
@alexeystrakh did you get it working with the `get-task-allow` entitlement?

@Eilon we also have https://github.com/dotnet/maui/issues/5561 which tracks enabling dev tools for Blazor WebView. Do we want to keep this issue open for MAUI WebView or close it as a duplicate?

**Comment by Eilon** [comment]
@TanayParikh ah I forgot we had a BlazorWebView-specific bug, so let's keep that, and then this bug will be for MAUI WebView.

That feedback ticket is unfortunately not visible even after I log in with my Apple ID.

**Comment by alexeystrakh** [comment]
I'm following [this guide](https://github.com/dotnet/docs-maui/blob/main/docs/ios/deployment/entitlements.md) and have added the `Entitlements.plist` with the required value and registered it within a property group property `CodeSignEntitlements`, although it the build logs I don't see it was used by the build:
 
```
Target "_CompileEntitlements" in file "/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.4.402/tools/msbuild/iOS/Xamarin.Shared.targets":
  Building target "_CompileEntitlements" completely.
  No input files were specified.
  Using "CompileEntitlements" task from assembly "/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.4.402/tools/msbuild/iOS/../iOS/Xamarin.iOS.Tasks.dll".
  Task "CompileEntitlements"
    /usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.4.402/tools/msbuild/iOS/Xamarin.Shared.targets(627,3): warning : Cannot expand $(AppIdentifierPrefix) in Entitlements.plist without a provisioning profile.
    /usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.4.402/tools/msbuild/iOS/Xamarin.Shared.targets(627,3): warning :         
  Done executing task "CompileEntitlements".
  Task "GetFullPath" skipped, due to false condition; ('$(IsMacEnabled)' == 'true' And '$(_CompiledEntitlements)' != '') was evaluated as ('true' == 'true' And '' != '').
  Using "GetFullPath" task from assembly "/usr/local/share/dotnet/packs/Microsoft.MacCatalyst.Sdk/15.4.402/tools/msbuild/iOS/../iOS/Xamarin.iOS.Tasks.dll".
  Task "GetFullPath"
  Done executing task "GetFullPath".
Done building target "_CompileEntitlements" in project "maui-app-webviewtest2.csproj".
```
 
I'm using a personal Apple Account and wondering how I can get a key+profile for this. When I do the same via Xcode, it does everything automatically and also (!) says that a provisioning profile is not required:
 
![image](https://user-images.githubusercontent.com/3697084/173142562-57001c9e-5b39-47f4-b903-cf5a2757bc9d.png)
 
Because of this, I'm unable to validate if the `Entitlement.plist` with the get-task-allow helps with the issue.

**Comment by rolfbjarne** [comment]
Could you zip up and attach the .app from Xcode (that works) and the .app from MAUI (that doesn't)?

**Comment by VincentBu** [comment]
repro with maui main branch on mac m1

**Comment by zengxs** [comment]
`get-task-allow` entitlement does make it work.

I added configuration like below in my `csproj` file:
~~~xml
<PropertyGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">
    <CodeSignEntitlements>Platforms/MacCatalyst/Entitlements.Debug.plist</CodeSignEntitlements>
</PropertyGroup>
~~~

The content of the `Entitlements.Debug.plist` file is as follows:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>com.apple.security.app-sandbox</key>
    <false/>
    <key>com.apple.security.get-task-allow</key>
    <true/>
</dict>
</plist>
```

Then rebuild and run the app, safari developer tools inspector work fine:

<img width="966" alt="image" src="https://user-images.githubusercontent.com/18748098/183805007-6de00019-ca05-4614-925c-2639f8a6457d.png">

**Comment by Redth** [comment]
@davidbritch is this something you could add to the conceptual docs?

@dalexsoto in theory we could make a short hand way to do this? Perhaps a property that adds an entitlements file to the build? @rolfbjarne is there a task that merges partial entitlements files like there is for info.plist?

**Comment by TanayParikh** [comment]
@Redth would adding this enlistment profile to the template be considered?

**Comment by davidbritch** [comment]
@Redth Yes, can add to the docs. I've created an issue for it.

**Comment by julianadormon** [comment]
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
> <plist version="1.0">
> <dict>
>     <key>com.apple.security.app-sandbox</key>
>     <false/>
>     <key>com.apple.security.get-task-allow</key>
>     <true/>
> </dict>
> </plist>
> ```

@zengxs suggestion did not work for me using Visual Studio Community 2022 for Mac Preview
Version 17.4 Preview (17.4 build 715)

**Comment by zengxs** [comment]
@julianadormon Sorry for the slow reply.

What is your Xcode and system version? My environment is macOS Monterey 12.5 on Apple Silicon, Xcode 13.4.1.
Do you have the conditions to try it on a new machine? I guess it might be a problem with the system environment.

**Comment by rolfbjarne** [comment]
> @rolfbjarne is there a task that merges partial entitlements files like there is for info.plist?

No, there isn't a way to add a partial entitlements file into the build right now.

**Comment by tstephansen** [comment]
This issue has apparently come back. I added the entitlements plist file and the necessary code to the csproj file. This worked before the update to Ventura 13.3 but it is no longer working after the update. Is there a workaround to use the developer tools in the latest version of OSX?

**Comment by drasticactions** [comment]
Trying with Xcode 14.2, On Catalyst, in a Swift App, with the included plist values and setting the developer flag, I can't get it working there either: https://github.com/drasticactions/MauiRepros/tree/main/WebKitTools

It works on iPhone, not Catalyst, and that's using Swift.

**Comment by drasticactions** [comment]
@tstephansen I believe this is unrelated to MAUI. 

https://github.com/drasticactions/MauiRepros/tree/main/WebViewChartTest

The developer tools don't work from a straight Swift or Objective-C Mac Catalyst app either. I think this was broken by Apple at some point between when this issue was "fixed" and today.

**Comment by tstephansen** [comment]
Yeah it coincided with the release of 13.3 I think (at least that's when it stopped working for me). Sorry about that! I didn't understand the previous comment (I thought they were trying to debug Maui in Xcode for some reason).

**Comment by cs-italy** [comment]
According to [this page](https://webkit.org/blog/13936/enabling-the-inspection-of-web-content-in-apps/), you can make a WKWebView inspectable like this:

```
#if DEBUG
#if MACCATALYST13_3_OR_GREATER
        webview.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
#elif IOS16_1_OR_GREATER
        if (DeviceInfo.Version.Minor >= 4)
            webview.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
#endif
#endif
```

**Comment by drasticactions** [comment]
https://developer.apple.com/documentation/webkit/wkwebview/4111163-isinspectable

Seems like it was added for 16.4

https://github.com/xamarin/xamarin-macios/blob/main/src/webkit.cs#L5507-L5509

It should be bound in the next Xamarin.MaciOS Xcode update. Trying out that selector you wrote in Swift and C# it worked with the older SDK.

<img width="349" alt="スクリーンショット 2023-04-14 17 10 54" src="https://user-images.githubusercontent.com/898335/231984826-3a7690ed-8067-4fce-ae5f-1a3bd8629942.png">

So you could use the selector for now until the next SDK bump.

**Comment by headintheclouds21** [comment]
Has anyone been able to get this working again? I've added the get-task-allow entitlement and followed instructions at https://learn.microsoft.com/en-us/dotnet/maui/user-interface/controls/webview?view=net-maui-7.0&pivots=devices-maccatalyst with no luck. I'm using macOS 12.6.3 with xcode 14.2. This is a large issue for me and being able to inspect a webview would go a long ways

**Comment by tstephansen** [comment]
@headintheclouds21 I don't have enough experience working with Xamarin/Maui to get it to work. I'm sure there's a way with the new property that @cs-italy replied with and I've just been debugging on Windows for the time being. It would be great if someone who does know how to implement this could give an example though.

**Comment by tstephansen** [comment]
@headintheclouds21 I think I got it figured out. I haven't actually tested and JS in it yet (busy on another project) but it at least shows up in Safari so I suspect JS debugging will work.

So what I did was create a project with the default maui blazor template. I also made the changes outlined in the PR that fixed this issue. The trick to getting this to work (for me at least) is adding the following to the MainPage.xaml.cs file.

```
public MainPage()
{
    InitializeComponent();
    Loaded += MainPage_Loaded;
}

private void MainPage_Loaded(object sender, EventArgs e)
{
    if (blazorWebView.Handler.PlatformView is WKWebView view)
    {
        view.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
    }
}
```

Once I did that the app shows up in Safari again.

![CleanShot 2023-05-03 at 13 48 48@2x](https://user-images.githubusercontent.com/8071398/236002037-7af92362-4cdc-4c39-a678-4144dbb7d6c2.png)

**Comment by wadie82** [comment]
> @headintheclouds21 I think I got it figured out. I haven't actually tested and JS in it yet (busy on another project) but it at least shows up in Safari so I suspect JS debugging will work.
> 
> So what I did was create a project with the default maui blazor template. I also made the changes outlined in the PR that fixed this issue. The trick to getting this to work (for me at least) is adding the following to the MainPage.xaml.cs file.
> 
> ```
> public MainPage()
> {
>     InitializeComponent();
>     Loaded += MainPage_Loaded;
> }
> 
> private void MainPage_Loaded(object sender, EventArgs e)
> {
>     if (blazorWebView.Handler.PlatformView is WKWebView view)
>     {
>         view.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
>     }
> }
> ```
> 
> Once I did that the app shows up in Safari again.
> 
> ![CleanShot 2023-05-03 at 13 48 48@2x](https://user-images.githubusercontent.com/8071398/236002037-7af92362-4cdc-4c39-a678-4144dbb7d6c2.png)

hello.  Can you please let me know what are the PR changes you refer to. The conversation of this thread is so hard to follow as they are way too many comments :) Apologise I could not find the full solution.

**Comment by drasticactions** [comment]
The issue linked to this one has nothing to do with what I fixed:

https://github.com/dotnet/maui/pull/14610

In short, Apple replaced the APIs for setting the debug flag to make the view inspectable. I changed the defaults for Blazor WebViews to enable it by default for debug mode, but you can call the same APIs right now in your app to enable it again, which is what @tstephansen did. 

Once the webview is loaded (either by checking for the Handler Change event, or when the page loads) you can invoke the handlers platform view and set the flags to enable it.

(Again, to make it clear to people in this thread, _this was never a MAUI bug, Apple changed the APIs and we changed the code to match_)

**Comment by wadie82** [comment]
sorry guys. I have tried it again. I run into issues such as NSObject is not defined neither NSString. I am definitely missing an import somehow for those. would you know how to resolve ?

**Comment by Eilon** [comment]
> sorry guys. I have tried it again. I run into issues such as NSObject is not defined neither NSString. I am definitely missing an import somehow for those. would you know how to resolve ?

Try this:

```c#
platformView.SetValueForKey(global::Foundation.NSObject.FromObject(enableWebDevTools), new global::Foundation.NSString("inspectable"));
```

**Comment by tstephansen** [comment]
> > sorry guys. I have tried it again. I run into issues such as NSObject is not defined neither NSString. I am definitely missing an import somehow for those. would you know how to resolve ?
> 
> Try this:
> 
> ```cs
> platformView.SetValueForKey(global::Foundation.NSObject.FromObject(enableWebDevTools), new global::Foundation.NSString("inspectable"));
> ```

If you're building for a platform other than MacCatalyst you might get that message. Try adding a conditional if the advice above didn't help.

```
public MainPage()
{
    InitializeComponent();
    Loaded += MainPage_Loaded;
}

private void MainPage_Loaded(object sender, EventArgs e)
{
#if DEBUG && MACCATALYST13_3_OR_GREATER
    if (blazorWebView.Handler.PlatformView is WKWebView view)
    {
        view.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
    }
#endif
}
```

**Comment by v-Lily** [comment]
@tstephansen 

Based on your comment, I tried adding the above code, Mac IOS simulator still doesn't allow in Safari developer tools.
![image](https://github.com/dotnet/maui/assets/83056958/82bca597-9a19-4a4c-a84e-63e0c7d1c797)

**Comment by tstephansen** [comment]
@v-Lily [I created an example repository](https://github.com/tstephansen/MauiBlazorInspectable) where inspectable has been implemented. Hope this helps!

**Comment by v-Lily** [comment]
> @v-Lily [I created an example repository](https://github.com/tstephansen/MauiBlazorInspectable) where inspectable has been implemented. Hope this helps!

Thanks for your comment.
According to your [example repository](https://github.com/tstephansen/MauiBlazorInspectable), the Mac IOS simulator is still not allowed in the Safari developer tools, but the MacCatalyst simulator is available in the Safari developer tools.
![image](https://github.com/dotnet/maui/assets/83056958/e7921a14-d132-4d03-a6ec-3ea323eec8bd)

**Comment by gerneio** [comment]
For me, the preprocessing directives weren't firing on my iOS Simulator device, so I switched it to this:

```csharp
    private partial void BlazorWebViewInitialized(object? sender, BlazorWebViewInitializedEventArgs e) 
    {
        // Enable Developer Extras for Catalyst/iOS builds for 16.4+
        if ((DeviceInfo.Current.Platform == DevicePlatform.MacCatalyst && DeviceInfo.Current.Version >= new Version(13, 3)) ||
            (DeviceInfo.Current.Platform == DevicePlatform.iOS && DeviceInfo.Current.Version >= new Version(16, 4)))
        {
            e.WebView.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
        }
    }
```

**Comment by abivelj** [comment]
> > @headintheclouds21 I think I got it figured out. I haven't actually tested and JS in it yet (busy on another project) but it at least shows up in Safari so I suspect JS debugging will work.
> > So what I did was create a project with the default maui blazor template. I also made the changes outlined in the PR that fixed this issue. The trick to getting this to work (for me at least) is adding the following to the MainPage.xaml.cs file.
> > ```
> > public MainPage()
> > {
> >     InitializeComponent();
> >     Loaded += MainPage_Loaded;
> > }
> > 
> > private void MainPage_Loaded(object sender, EventArgs e)
> > {
> >     if (blazorWebView.Handler.PlatformView is WKWebView view)
> >     {
> >         view.SetValueForKey(NSObject.FromObject(true), new NSString("inspectable"));
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Once I did that the app shows up in Safari again.
> > ![CleanShot 2023-05-03 at 13 48 48@2x](https://user-images.githubusercontent.com/8071398/236002037-7af92362-4cdc-4c39-a678-4144dbb7d6c2.png)
> 
> hello. Can you please let me know what are the PR changes you refer to. The conversation of this thread is so hard to follow as they are way too many comments :) Apologise I could not find the full solution.

I followed what you said, but then changed it slightly.

`#if MACCATALYST
        WebKit.WKWebView view = (WebKit.WKWebView)blazorWebView.Handler.PlatformView;
        view.SetValueForKey(Foundation.NSObject.FromObject(true) , new Foundation.NSString("inspectable"));
#endif`
